src/analysis/ring_seminorm.lean

```diff
@@ -0,0 +1,188 @@
+
+instance : inhabited (ring_seminorm R) := ⟨0⟩
+
+variables (p : ring_seminorm R)
+
+protected lemma nonneg : 0 ≤ p x := p.nonneg' _
+@[simp] protected lemma map_zero : p 0 = 0 := p.map_zero'
+protected lemma add_le : p (x + y) ≤ p x + p y := p.add_le' _ _
+@[simp] protected lemma neg : p (- x) = p x := p.neg' _
+protected lemma mul_le : p (x * y) ≤ p x * p y := p.mul_le' _ _
+
+lemma pow_le : ∀ {n : ℕ}, 0 < n → p (x ^ n) ≤ (p x) ^ n
+| 1 h := by simp only [pow_one]
+| (n + 2) h :=
+by simpa only [pow_succ _ (n + 1)] using le_trans (p.mul_le x _)
+  (mul_le_mul_of_nonneg_left (pow_le n.succ_pos) (p.nonneg _))
+
+/-- A function `f : R → ℝ` from a ring `R` satisfies `is_norm_le_one_class` if `f 1 ≤ 1`. -/
+def is_norm_le_one_class {R : Type*} [semiring R] (f : R → ℝ) : Prop := f 1 ≤ 1
+
+/-- A function `f : R → ℝ` from a ring `R` satisfies `is_norm_one_class` if `f 1 = 1`. -/
+def is_norm_one_class {R : Type*} [semiring R] (f : R → ℝ) : Prop := f 1 = 1
+
+lemma is_norm_one_class_iff_nonzero (hp : is_norm_le_one_class p) :
+  is_norm_one_class p ↔ ∃ x : R, p x ≠ 0 :=
+begin
+  rw is_norm_one_class,
+  refine ⟨λ h, ⟨1, by {rw h, exact one_ne_zero}⟩, λ h, _⟩,
+  { obtain ⟨x, hx⟩ := h,
+    by_cases hp0 : p 1 = 0,
+    { have hx' : p x ≤ 0,
+      { rw ← mul_one x,
+        apply le_trans (p.mul_le x 1) _,
+        rw [hp0, mul_zero], },
+      exact absurd (le_antisymm hx' (p.nonneg x) ) hx },
+    { have h1 : p 1 * 1 ≤ p 1 * p 1,
+      { conv_lhs { rw ← one_mul (1 : R) },
+        convert p.mul_le 1 1,
+        rw mul_one, },
+      rw mul_le_mul_left (lt_of_le_of_ne (p.nonneg _) (ne.symm hp0)) at h1,
+      exact le_antisymm hp h1, }}
+end
+
+end ring_seminorm
+
+/-- The norm of a semi_normed_ring as a ring_seminorm. -/
+def norm_ring_seminorm (R : Type*) [semi_normed_ring R] :
+  ring_seminorm R :=
+{ to_fun    := norm,
+  map_zero' := norm_zero,
+  nonneg'   := norm_nonneg,
+  add_le'   := norm_add_le,
+  neg'      := norm_neg,
+  mul_le'   := norm_mul_le }
+
+/-- A function `f : R → ℝ` is a norm if it is a seminorm and `f x = 0` implies `x = 0`. -/
+structure ring_norm (R : Type*) [ring R] extends (ring_seminorm R) :=
+(ne_zero : ∀ x, x ≠ 0 → 0 < to_fun x)
+
+attribute [nolint doc_blame] ring_norm.to_ring_seminorm
+
+namespace ring_norm
+
+variable [ring R]
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : has_coe_to_fun (ring_norm R) (λ _, R → ℝ) := ⟨λ p, p.to_fun⟩
+
+instance zero_hom_class : zero_hom_class (ring_norm R) R ℝ :=
+ { coe := λ f, f.to_fun,
+   coe_injective' := λ f g h, by cases f; cases g; congr',
+   map_zero := λ f, f.map_zero' }
+
+@[ext] lemma ext {p q : ring_norm R} (h : ∀ x, p x = q x) : p = q :=
+fun_like.ext p q h
+
+variable (R)
+
+/-- The trivial norm on a ring `R` is the `ring_norm` taking value `0` at `0` and `1` at every
+  other element. -/
+def trivial_norm [decidable_eq R] : ring_norm R :=
+{ to_fun    := λ x, if x = 0 then 0 else 1,
```

Comment:
Can you define this for `add_group_seminorm`s too?

Reply:
I did define it, but I'm not sure how useful it will be in that setting.